`copilot-theorem`: Examples do not compile

**Description**

Some example programs in the `copilot-theorem/example` subdirectory have bitrotted, as they depend on a `Copilot.Theorem.Prover.Z3` module that is no longer included in `copilot-theorem.cabal` (as of https://github.com/Copilot-Language/copilot/commit/6a67d89ee42ffa0718e844e03a4e7f34d2a88801). Attempting to build any of these examples results in compilation failures.

**Type**

* Bug: Example programs fail to compile.

**Additional context**

`Copilot.Theorem.Prover.Z3` was an early, SMT solver–based approach for proving properties about Copilot specifications. The `copilot-theorem` library has largely replaced `Copilot.Theorem.Prover.Z3` will different APIs, such as those offered by `Copilot.Theorem.Prover.SMT` and `Copilot.Theorem.What4`. Of the latter two, the API in `Copilot.Theorem.Prover.SMT` is closest to what `Copilot.Theorem.Prover.Z3` offered.

**Requester**

* Ryan Scott (Galois)

**Method to check presence of bug**

First, build the `copilot` libraries, using a method such as the following:

```
$ cabal build copilot --write-ghc-environment-files=always
```

Then invoke `ghc` on one of the files in the `copilot-theorem/examples` subdirectory that import `Copilot.Theorem.Prover.Z3`. The bug manifests as a GHC compilation error, such as in the following example:

```
$ ghc copilot-theorem/examples/BoyerMoore.hs
Loaded package environment from /Users/rscott/Documents/Hacking/Haskell/copilot/.ghc.environment.aarch64-darwin-9.4.8
[1 of 1] Compiling BoyerMoore       ( copilot-theorem/examples/BoyerMoore.hs, copilot-theorem/examples/BoyerMoore.o )

copilot-theorem/examples/BoyerMoore.hs:8:1: error:
    Could not find module ‘Copilot.Theorem.Prover.Z3’
    Perhaps you meant
      Copilot.Theorem.Prover.SMT (from copilot-theorem-4.6)
      Copilot.Theorem.Prover.SMT (needs flag -package-id copilot-theorem-4.3)
      Copilot.Theorem.Prover.SMT (needs flag -package-id copilot-theorem-4.3)
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
8 | import Copilot.Theorem.Prover.Z3
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Expected result**

Invoking `ghc` on any of the examples in the `copilot-theorem/examples` subdirectory should succeed.

**Desired result**

Invoking `ghc` on any of the examples in the `copilot-theorem/examples` subdirectory should succeed.

**Proposed solution**

Modify the code in each of the examples in `copilot-theorem/examples` to import from `Copilot.Theorem.Prover.SMT` instead of `Copilot.Theorem.Prover.Z3`, making adjustments as needed to adapt to the slightly different API in `Copilot.Theorem.Prover.SMT`.

**Further notes**

None.

--- a/copilot-theorem/CHANGELOG
+++ b/copilot-theorem/CHANGELOG
@@ -1,3 +1,6 @@
+2026-03-31
+        * Fix examples so they compile with latest copilot-theorem. (#692)
+
 2026-03-07
         * Version bump (4.7). (#714)
 

--- a/copilot-theorem/CHANGELOG
+++ b/copilot-theorem/CHANGELOG
@@ -1,5 +1,6 @@
 2026-03-31
         * Handle special Float values correctly for counterexamples. (#697)
+        * Update examples to compile with copilot-theorem >= 3.0. (#692)
 
 2026-03-07
         * Version bump (4.7). (#714)

--- a/copilot-theorem/examples/BoyerMoore.hs
+++ b/copilot-theorem/examples/BoyerMoore.hs
@@ -5,7 +5,7 @@ module BoyerMoore where
 
 import Copilot.Language hiding (length)
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (induction, def, debug, z3)
 
 import Copilot.Core.Type
 
@@ -71,7 +71,7 @@ spec = do
     maj = majorityVote ss
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 -- | Initial value for a given type.
 --

--- a/copilot-theorem/examples/BoyerMoore.hs
+++ b/copilot-theorem/examples/BoyerMoore.hs
@@ -5,7 +5,7 @@ module BoyerMoore where
 
 import Copilot.Language hiding (length)
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (debug, def, induction, z3)
 
 import Copilot.Core.Type
 
@@ -71,7 +71,7 @@ spec = do
     maj = majorityVote ss
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 -- | Initial value for a given type.
 --

--- a/copilot-theorem/examples/Grey.hs
+++ b/copilot-theorem/examples/Grey.hs
@@ -4,7 +4,7 @@ module Grey where
 
 import Copilot.Language
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (def, Options (..), z3, kInduction, induction)
 
 import Prelude ()
 import Data.String (fromString)
@@ -32,7 +32,7 @@ spec = do
     r  = extern "reset" Nothing
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 kinduct :: Word32 -> Proof Universal
-kinduct k = kInduction def { nraNLSat = False, startK = k, maxK = k, debug = False }
+kinduct k = kInduction def { startK = k, maxK = k, debug = False } z3

--- a/copilot-theorem/examples/Grey.hs
+++ b/copilot-theorem/examples/Grey.hs
@@ -4,7 +4,7 @@ module Grey where
 
 import Copilot.Language
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (Options (..), def, induction, kInduction, z3)
 
 import Prelude ()
 import Data.String (fromString)
@@ -32,7 +32,7 @@ spec = do
     r  = extern "reset" Nothing
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 kinduct :: Word32 -> Proof Universal
-kinduct k = kInduction def { nraNLSat = False, startK = k, maxK = k, debug = False }
+kinduct k = kInduction def { startK = k, maxK = k, debug = False } z3

--- a/copilot-theorem/examples/Incr.hs
+++ b/copilot-theorem/examples/Incr.hs
@@ -4,7 +4,7 @@ import Prelude ()
 import Copilot.Language
 
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (def, induction, debug, z3)
 
 spec = do
   bounds <- prop "bounds" (forAll $ x < 255)
@@ -16,4 +16,4 @@ spec = do
     x = [2] ++ (1 + x)
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = True }
+induct = induction def { debug = True } z3

--- a/copilot-theorem/examples/Incr.hs
+++ b/copilot-theorem/examples/Incr.hs
@@ -4,7 +4,7 @@ import Prelude ()
 import Copilot.Language
 
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (debug, def, induction, z3)
 
 spec = do
   bounds <- prop "bounds" (forAll $ x < 255)
@@ -16,4 +16,4 @@ spec = do
     x = [2] ++ (1 + x)
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = True }
+induct = induction def { debug = True } z3

--- a/copilot-theorem/examples/SerialBoyerMoore.hs
+++ b/copilot-theorem/examples/SerialBoyerMoore.hs
@@ -4,7 +4,7 @@ module SerialBoyerMoore where
 
 import Copilot.Language
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (def, Options (..), induction, kInduction, z3)
 
 import Prelude ()
 import Data.String (fromString)
@@ -56,7 +56,7 @@ spec = do
     (p, s, j) = majority input
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 kinduct :: Word32 -> Proof Universal
-kinduct k = kInduction def { nraNLSat = False, startK = k, maxK = k, debug = False }
+kinduct k = kInduction def { startK = k, maxK = k, debug = False } z3

--- a/copilot-theorem/examples/SerialBoyerMoore.hs
+++ b/copilot-theorem/examples/SerialBoyerMoore.hs
@@ -4,7 +4,7 @@ module SerialBoyerMoore where
 
 import Copilot.Language
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (Options (..), def, induction, kInduction, z3)
 
 import Prelude ()
 import Data.String (fromString)
@@ -56,7 +56,7 @@ spec = do
     (p, s, j) = majority input
 
 induct :: Proof Universal
-induct = induction def { nraNLSat = False, debug = False }
+induct = induction def { debug = False } z3
 
 kinduct :: Word32 -> Proof Universal
-kinduct k = kInduction def { nraNLSat = False, startK = k, maxK = k, debug = False }
+kinduct k = kInduction def { startK = k, maxK = k, debug = False } z3

--- a/copilot-theorem/examples/SphericalWCV.hs
+++ b/copilot-theorem/examples/SphericalWCV.hs
@@ -8,7 +8,7 @@ import Prelude ()
 import Copilot.Language
 import Copilot.Language.Reify
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (def, debug, onlySat, onlyValidity, z3)
 
 import qualified Copilot.Language.Operators.Propositional as P
 
@@ -241,7 +241,7 @@ localConvexity = do
   theorem "6"  (P.not (forAll $ locallyConvex tau)) arithSat
 
 arith :: Proof Universal
-arith    = onlyValidity def { nraNLSat = True, debug = False }
+arith    = onlyValidity def { debug = False } z3
 
 arithSat :: Proof Existential
-arithSat = onlySat      def { nraNLSat = True, debug = False }
+arithSat = onlySat      def { debug = False } z3

--- a/copilot-theorem/examples/SphericalWCV.hs
+++ b/copilot-theorem/examples/SphericalWCV.hs
@@ -8,7 +8,7 @@ import Prelude ()
 import Copilot.Language
 import Copilot.Language.Reify
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (debug, def, onlySat, onlyValidity, z3)
 
 import qualified Copilot.Language.Operators.Propositional as P
 
@@ -241,7 +241,7 @@ localConvexity = do
   theorem "6"  (P.not (forAll $ locallyConvex tau)) arithSat
 
 arith :: Proof Universal
-arith    = onlyValidity def { nraNLSat = True, debug = False }
+arith    = onlyValidity def { debug = False } z3
 
 arithSat :: Proof Existential
-arithSat = onlySat      def { nraNLSat = True, debug = False }
+arithSat = onlySat      def { debug = False } z3

--- a/copilot-theorem/examples/WCV.hs
+++ b/copilot-theorem/examples/WCV.hs
@@ -5,7 +5,7 @@ import Prelude ()
 import Copilot.Language
 import Copilot.Language.Reify
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (def, debug, onlySat, onlyValidity, z3)
 
 import qualified Copilot.Language.Operators.Propositional as P
 
@@ -165,7 +165,7 @@ localConvexity = do
   theorem "6"  (P.not (forAll $ locallyConvex tau)) arithSat
 
 arith :: Proof Universal
-arith    = onlyValidity def { nraNLSat = True, debug = False }
+arith    = onlyValidity def { debug = False } z3
 
 arithSat :: Proof Existential
-arithSat = onlySat      def { nraNLSat = True, debug = False }
+arithSat = onlySat      def { debug = False } z3

--- a/copilot-theorem/examples/WCV.hs
+++ b/copilot-theorem/examples/WCV.hs
@@ -5,7 +5,7 @@ import Prelude ()
 import Copilot.Language
 import Copilot.Language.Reify
 import Copilot.Theorem
-import Copilot.Theorem.Prover.Z3
+import Copilot.Theorem.Prover.SMT (debug, def, onlySat, onlyValidity, z3)
 
 import qualified Copilot.Language.Operators.Propositional as P
 
@@ -165,7 +165,7 @@ localConvexity = do
   theorem "6"  (P.not (forAll $ locallyConvex tau)) arithSat
 
 arith :: Proof Universal
-arith    = onlyValidity def { nraNLSat = True, debug = False }
+arith    = onlyValidity def { debug = False } z3
 
 arithSat :: Proof Existential
-arithSat = onlySat      def { nraNLSat = True, debug = False }
+arithSat = onlySat      def { debug = False } z3

